Ignore Content-Length from env.request_headers

pilaf opened the pull request:
Closes #51

This makes the adapter ignore `Content-Length` if given in Faraday's `env.request_headers`, but only if capitalized, since `Async::HTTP` will set `content-length` anyway.

I'd like to add tests too, but I wasn't sure how to test what headers end up in the final request without setting up a dummy server to receive them. Any suggestions?

ioquatix commented on lib/async/http/faraday/adapter.rb:
```diff
@@ -181,6 +181,12 @@ def perform_request(env)
 						end
 						
 						if headers = env.request_headers
+              # Ignore Content-Length if given, it will be set for us later anyway (lowercased)
```
This is looking good.

I'd suggest extracting this into `length` and then assigning to `BodyReadWrapper`, which should have this:

```
class BodyReadWrapper
  def initialize(body, length = nil, block_size: 4096)
    @body = body
    @length = length
    @block_size = block_size
  end

  # ...

  def length
    @length # if known
  end
```

You'll need to move the `if headers = env.request_headers` first to extract a local variable if possible.

Is `headers.key?(...)` case sensitive?

Testing is a bit tricky, but maybe we can mock the client. If you can get the code updated, I can help write the tests.

pilaf replied:
@ioquatix 

> Is headers.key?(...) case sensitive?

Yes. I'd change this to check for both `Content-Length` and `content-length` if the intent is to use that value for `BodyReadWrapper#length`, if you agree.

pilaf commented on lib/async/http/faraday/adapter.rb:
```diff
@@ -168,22 +173,29 @@ def call(env)
+								body = BodyReadWrapper.new(body, content_length)
 							else
-								body = ::Protocol::HTTP::Body::Buffered.wrap(body)
+								body = ::Protocol::HTTP::Body::Buffered.wrap(body).then do |buffered|
+									::Protocol::HTTP::Body::Buffered.new(buffered.chunks, content_length)
```
This double initialization is ugly but I had no choice since `Protocol::HTTP::Body::Buffered.wrap` doesn't take a `length` argument.

ioquatix replied:
I understand how you are thinking about the problem.

A buffered body can compute its size accurately and efficiently from the chunks. That's different from a readable body where we can only call `#read` to stream the data.

Therefore, I wouldn't worry about this code path - just leave it as the original implementation.

pilaf commented on lib/async/http/faraday/adapter.rb:
```diff
@@ -168,22 +173,29 @@ def call(env)
 						if body = env.body
 							# We need to ensure the body is wrapped in a Readable object so that it can be read in chunks:
 							# Faraday's body only responds to `#read`.
 							if body.is_a?(::Protocol::HTTP::Body::Readable)
 								# Good to go
```
Should this case try to rebuild the body with the given length too?

ioquatix replied:
No I wouldn’t worry about this case either, it’s expected the body would know its size already if appropriate.

The main issue we are addressing is Faradays opaque “body#read” which may have length but the only way to know that is via the supplied header.

pilaf commented:
@ioquatix Sorry for the delay, please check the latest commit, I think it mostly addresses your comments.

Again I couldn't write tests for it, but I did test that this code works separately.

ioquatix commented:
Thanks, this looks like what I was thinking of, I'll try to add some tests, and if it looks good we should be able to move forward.

ioquatix commented on lib/async/http/faraday/adapter.rb:
```diff
@@ -168,22 +173,29 @@ def call(env)
+							# Use content-length to inform body length if given, but remove the header since it will be
+							# set for us later anyway, and not doing so could result in a duplicate content-length headers
+							# if capitalization differs
+							content_length = headers.delete("content-length")&.to_i
```
This will give a content length of zero if the header was not set, e.g. `nil.to_i == 0`. I think it's better to write it like this:

```suggestion
							if value = headers.delete("content-length")
							  content_length = value.to_i
							end
```

pilaf replied:
It will not be set to zero because of `&.`, which prevents calling `.to_i` if the callee is `nil`:

```
irb(main):001> nil&.to_i
=> nil
```

But if you think that syntax is unclear I can change it.

ioquatix commented:
This is perfect, thanks.

balvig commented:
I ran into this issue myself today, so just came here to say thanks in advance 😅 and let me know if I can help with anything to move things forward! 🙏

ioquatix commented:
I made it an error to specify singleton headers more than once, and now it fails:

```
> bundle exec sus
  0.0s     warn: Async::Task: Reading HTTP/1.1 requests for Async::HTTP::Protocol::HTTP1::Server. [oid=0x718] [ec=0x720] [pid=981323] [2025-07-01 21:35:34 +1200]
               | Task may have ended with unhandled exception.
               |   Protocol::HTTP::DuplicateHeaderError: Duplicate singleton header key: "content-length"
               |   → /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:336 in 'Protocol::HTTP::Headers#merge_into'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:356 in 'block in Protocol::HTTP::Headers#to_h'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:355 in 'Array#each'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:355 in 'Enumerable#inject'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:355 in 'Protocol::HTTP::Headers#to_h'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http-0.51.0/lib/protocol/http/headers.rb:348 in 'Protocol::HTTP::Headers#[]'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http1-0.34.0/lib/protocol/http1/connection.rb:169 in 'Protocol::HTTP1::Connection#persistent?'
               |     /home/samuel/.gem/ruby/3.4.1/gems/protocol-http1-0.34.0/lib/protocol/http1/connection.rb:401 in 'Protocol::HTTP1::Connection#read_request'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-http-0.89.0/lib/async/http/protocol/http1/request.rb:26 in 'Async::HTTP::Protocol::HTTP1::Request.read'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-http-0.89.0/lib/async/http/protocol/http1/server.rb:49 in 'Async::HTTP::Protocol::HTTP1::Server#next_request'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-http-0.89.0/lib/async/http/protocol/http1/server.rb:66 in 'Async::HTTP::Protocol::HTTP1::Server#each'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-http-0.89.0/lib/async/http/server.rb:50 in 'Async::HTTP::Server#accept'
               |     /home/samuel/.gem/ruby/3.4.1/gems/io-endpoint-0.15.2/lib/io/endpoint/wrapper.rb:216 in 'block (2 levels) in IO::Endpoint::Wrapper#accept'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-2.25.0/lib/async/task.rb:201 in 'block in Async::Task#run'
               |     /home/samuel/.gem/ruby/3.4.1/gems/async-2.25.0/lib/async/task.rb:439 in 'block in Async::Task#schedule'
24 passed 1 errored out of 25 total (132 assertions)
🏁 Finished in 2.7s; 49.787 assertions per second.
🐢 Slow tests:
	992.7ms: describe Async::HTTP::Faraday::Adapter with a remote http server it can use a multi-part post body test/async/http/faraday/adapter.rb:270
	559.9ms: describe Async::HTTP::Faraday::Adapter with a remote http server it works without top level reactor test/async/http/faraday/adapter.rb:256
	472.2ms: describe Async::HTTP::Faraday::Adapter with a remote http server it works without initial url and trailing slash (compatiblisity to the original behaviour) test/async/http/faraday/adapter.rb:262
	409.7ms: describe Async::HTTP::Faraday::Adapter with a remote http server it can get remote resource test/async/http/faraday/adapter.rb:248
	103.8ms: describe Async::HTTP::Faraday::Adapter with a local http server with a slow response it client can get resource test/async/http/faraday/adapter.rb:145

🔥 Errored assertions:
describe Async::HTTP::Faraday::Adapter with a local http server it client can post resource test/async/http/faraday/adapter/post.rb:23
	⚠ Faraday::ConnectionFailed: EOFError
```
